### PR TITLE
Add distinction between (im)perfect recall in perfect information obs…

### PR DESCRIPTION
**Pull request by michalsustr**
…ervers.

### CODE REVIEW

File: open_spiel/observer.cc [hunk 1]
```diff
@@ -118,15 +118,26 @@ std::shared_ptr<Observer> Game::MakeObserver(
...
+  if (GetType().information == GameType::Information::kPerfectInformation) {
+    // Handle the dummy case, where we do not use any public information.
+    // The game will just have empty private observations.
+    if (!iig_obs_type->public_info) {
       return absl::make_unique<NoPrivateObserver>(*this);
     }
+
+    // Distinguish two variants:
+    // Provide observer with or without perfect recall.
+    SPIEL_CHECK_TRUE(iig_obs_type->public_info);
```
Comment (elkhrt) [inline open_spiel/observer.cc:130]: Remove this. Perhaps use an else clause.

File: open_spiel/observer.cc [hunk 2]
```diff
@@ -118,15 +118,26 @@ std::shared_ptr<Observer> Game::MakeObserver(
 
   // Perfect information games provide public information regardless
```
Comment (elkhrt) [inline open_spiel/observer.cc:119]: This comment is no longer accurate.

File: open_spiel/observer.cc (hunk 1 above)
Comment (elkhrt) [reply open_spiel/observer.cc:130]: Actually maybe this logic can be further simplified.

const bool perfect_info_game = GetType().information == GameType::Information::kPerfectInformation;
  if (perfect_info_game && !iig_obs_type->public_info) {
      return absl::make_unique<NoPrivateObserver>(*this);
 }

if (iig_obs_type->perfect_recall && (game_type_.provides_information_state_tensor || game_type_.provides_information_state_string)) {
  SPIEL_CHECK(perfect_info_game || iig_obs_type->private_info == PrivateInfoType::kSinglePlayer);
  return absl::make_unique<InformationStateObserver>(*this);
}

/ same for default observer

Comment (michalsustr) [reply open_spiel/observer.cc]: I simplified the logic.

File: open_spiel/observer.cc (hunk 2 above)
Comment (michalsustr) [reply open_spiel/observer.cc]: Why?

**Comment by elkhrt** [comment]
Looks like this affects some playthroughs. Is that deliberate?

**Comment by michalsustr** [comment]
I added the forgotten playthroughs. Now the perfect information games also log the public observations in playthroughs. It is redundant information, but I think it's ok for the tests.

File: open_spiel/observer.cc (hunk 2 above)
Comment (elkhrt) [reply open_spiel/observer.cc]: Because NoPrivateObserver does not provide public information.

**Pull request merged**